### ISSUE TITLE
setpersonality() seccomp stuff currently breaks CI

### ISSUE THREAD

**Issue by poettering** [systemd/systemd#6737]
Something is wrong with opinionated_personality(). Most likely the CI systems have personality set up weirdly, in a way opinionated_personality() didn't expect it.

**Comment by t8d7** [comment]
I'll work on this next weekend if someone doesn't already fix it by then.

**Comment by t8d7** [comment]
Previous discussion in #6730 (for tracking purposes)

**Comment by martinpitt** [comment]
This reproduces on Fedora 26 x86_64 in an i386 mock (`mock -r fedora-26-i386`). For the record, the CI runnners use a VM with a real i386 kernel, so the test seems to fail in both cases.

**Comment by evverx** [comment]
Judging by [how the `personality` wrapper works](https://sourceware.org/git/?p=glibc.git;a=commitdiff;h=e0043e17dfc52fe1702746543127cb4a87232bcd), I think it would be sufficient to use the following
```
assert_se(personality(...) == -EPERM);
```
instead of
```
errno = EUCLEAN;
assert_se(personality(...) == -1 && errno == EPERM);
```
when the test is run on `i386`. Or all the asserts can be replaced by `assert_se(personality(....) < 0)`.

**Comment by evverx** [comment]
I've just sent https://github.com/systemd/systemd/pull/6752.

By the way, could anyone clarify what the point of blocking either `PER_LINUX` or `PER_LINUX32` is? Why can't both be allowed?

**Comment by evverx** [comment]
I'm not entirely sure why `LockPersonality` currently blocks `0xffffffff`. If there is some reason for that, then it should probably be documented.

**Comment by topimiettinen** [comment]
@evverx the whole idea for `LockPersonality` is to disallow change of execution domain or personality. Also currently libseccomp would not support the logic needed for whitelisting multiple items while blacklisting the rest. It would be doable with BPF.

0xffffffff can be used to query the current personality. It shouldn't be blocked, but the same libseccomp limitation prevents this.

**Comment by topimiettinen** [comment]
@poettering perhaps opinionated_personality() should be just removed and current personality should be used, whatever it happens to be?

**Comment by evverx** [comment]
Docker allows using `PER_LINUX`, `PER_LINUX32`, `UNAME26` and `0xffffffff`, so it seems that some people might assume that `systemd` works in the same way. To prevent mix-ups, the documentation should probably make it clear what exactly `LockPersonality` locks.

**Comment by poettering** [comment]
> By the way, could anyone clarify what the point of blocking either PER_LINUX or PER_LINUX32 is? Why can't both be allowed?

On 32bit kernels we should probably permit either. on 64bit kernels it's how 32bit and 64bit environments are distuingished.

**Comment by evverx** [comment]
I still don't grasp what might be wrong with `PER_LINUX32` on 64bit kernels, but permitting `PER_LINUX` and `PER_LINUX32` on 32bit kernels seems to be reasonable.

**Comment by topimiettinen** [comment]
@evverx Maybe it's just my poor imagination, but I can't imagine any reason why any service should need to switch personalities. If a 32 bit service starts with `PER_LINUX` (or `PER_LINUX32`), that's fine and we should probably allow either, but why should a service need to switch to the other during its execution? But if such a service did indeed exist, it would be simple not to use `LockPersonality` in that case. 

The threats behind personalities could be that the emulations may be incomplete and buggy (fictive example: switch to 64 bit mode, mmap() something crossing 4 gigs limit, switch to 32 bit mode and back). In case of 32 bit in 64 bit system, system files for the secondary architecture may be less up to date than the primary architecture files, providing an easier attack surface than the primary environment. So I would not enable `PER_LINUX32` on 64 bit (even if it were possible with libseccomp). Again, if I'm wrong and something is out there which needs `PER_LINUX32`, it's a valid reason to avoid `LockPersonality` there.

**Comment by evverx** [comment]
@topimiettinen, I agree with you to some extent, but I think I need to read https://github.com/moby/moby/issues/20634 and https://github.com/moby/moby/issues/32839 thoroughly to find out why `docker` allows using `PER_LINUX32` and `UNAME26`.

Talking of `UNAME26`, the code which deals with it looks like the following:
```c

        if (current->personality & UNAME26) {
                const char *rest = UTS_RELEASE;
                char buf[65] = { 0 };
                int ndots = 0;
                unsigned v;
                size_t copy;

                while (*rest) {
                        if (*rest == '.' && ++ndots >= 3)
                                break;
                        if (!isdigit(*rest) && *rest != '.')
                                break;
                        rest++;
                }
                v = ((LINUX_VERSION_CODE >> 8) & 0xff) + 60;
                copy = clamp_t(size_t, len, 1, sizeof(buf));
                copy = scnprintf(buf, copy, "2.6.%u%s", v, rest);
                ret = copy_to_user(release, buf, copy + 1);
        }
        return ret;
```
and, to be honest, I don't see any reason why this has to be blocked by `LockPersonality`.

**Comment by topimiettinen** [comment]
I think Docker blocks personality() entirely.

The kernel part seems harmless, but user space libraries may change their behavior based on the uname, that's why this exists. Those compatibility code paths are probably not getting the same kind of attention as the other code base.

Wouldn't you be alarmed if logs revealed that systemd-timesyncd or wpa_supplicant had suddenly tried to switch the personality in any way? What could be the legitimate reason for personality switching for any service?

**Comment by evverx** [comment]
I'm not sure why you think that Docker blocks personality() entirely, because `PER_LINUX`, `PER_LINUX32`, `PER_LINUX|UNAME26`, `PER_LINUX32|UNAME26` and `0xffffffff` are whitelisted in
https://github.com/docker/docker/blob/be97c66708c24727836a22247319ff2943d91a03/profiles/seccomp/seccomp_default.go#L359-L413.

> Wouldn't you be alarmed if logs revealed that systemd-timesyncd or wpa_supplicant had suddenly tried to switch the personality in any way?

I would be alarmed. Now I'm wondering what the point of allowing using `PER_LINUX` is.

**Comment by evverx** [comment]
Given that the current implementation of `LockPersonality=` can deal with only one flag, would it make more sense to whitelist `0xffffffff` and blacklist everything else?

**Comment by topimiettinen** [comment]
@evverx You are right about Docker, I didn't read the code.

0xffffffff would make sense, but in my straces the only type of call of personality() wasn't that but PER_LINUX.

**Comment by evverx** [comment]
Strictly speaking, `personality(0)` is an attempt to change the kernel execution domain and `LockPersonality` is supposed to prevent such an attempt if it is not safe. `systemd` doesn't seem to be able to detect what is safe and what is not, so it appears to be safer to disallow everything except `0xffffffff`.

> 0xffffffff would make sense, but in my straces the only type of call of personality() wasn't that but PER_LINUX

What happens to the processes you were tracing when `personality(0)` fails?

**Comment by topimiettinen** [comment]
In my earlier tests I blocked `personality()` entirely with `SystemCallFilter` for all services and also for the DM, thus propagating to user programs. All services and almost all programs are completely unaffected (including Emacs 25), but here's a strace of Pulseaudio:
```
futex(0x7fdfeddfe8a8, FUTEX_WAKE_PRIVATE, 2147483647) = 0
personality(PER_LINUX <unfinished ...>) = ?
+++ killed by SIGSYS +++
Bad system call
```
Earlier I thought that some Python component used by pysolfc had issues of its own, but it's also hit by Pulseaudio.

**Comment by evverx** [comment]
I think that sending `SIGSYS` is quite a brutal way to block `personality` :-)

Pulseaudio calls [personality](https://github.com/pulseaudio/pulseaudio/blob/2417305ae755cbb3a92ca43a058f550809069cd9/src/pulsecore/core-util.c#L3426-L3433) and simply ignores any errors.

Emacs also calls `personality` if the `HAVE_PERSONALITY_ADDR_NO_RANDOMIZE` flag has been set when Emacs was compiled. As far as I understand Emacs ignores any errors too.

Basically, I'm trying to say that it would probably be better to start with allowing `0xffffffff`, which in my opinion should always be allowed, and add some other magic as soon as the current limitations of `libseccomp` can be got around.

**Comment by topimiettinen** [comment]
Yes, that makes sense.

**Comment by evverx** [comment]
@poettering, @tixxdz is there any chance you could say what you think of it? The patch allowing `0xffffffff` and disabling everything else seems to be easy to implement, but it would be great to know if such a patch would be accepted so as not to waste time implementing something doomed to be rejected :-)

**Comment by topimiettinen** [comment]
I opened an RFE for libseccomp for range checks (https://github.com/seccomp/libseccomp/issues/94).
